packages/remix-react/components.tsx

```diff
@@ -1296,7 +1296,45 @@ export function useMatches() {
  *
  * @see https://remix.run/api/remix#useloaderdata
  */
-export function useLoaderData<T = AppData>(): T {
+
+export type TypedResponse<T> = Response & {
+  json(): Promise<T>;
+};
+
+type DataFunction = (...args: any[]) => any; // matches any function
+type DataOrFunction = AppData | DataFunction;
+type JsonPrimitives = string | number | boolean | null;
+type NonJsonPrimitives = undefined | Function | symbol;
+type SerializeType<T> = T extends JsonPrimitives
+  ? T
+  : T extends undefined
+  ? undefined
+  : T extends { toJSON(): infer U }
+  ? U
+  : T extends []
+  ? []
+  : T extends [any, ...any[]]
+  ? {
+      [k in keyof T]: T[k] extends NonJsonPrimitives
+        ? null
+        : SerializeType<T[k]>;
+    }
+  : T extends (infer U)[]
+  ? SerializeType<U>[]
+  : {
+      [k in keyof T as T[k] extends NonJsonPrimitives
+        ? never
+        : k]: SerializeType<T[k]>;
+    };
+
+export type UseDataFunctionReturn<T extends DataOrFunction> = T extends (
+  ...args: any[]
```

Comment:
```suggestion
  ...args: unknown[]
```

Reply:
I've changed all the other `any`s  to `unknown` except this one. 

This specific change would break the code because TypeScript correctly treats function types are [contravariant](https://www.stephanboyer.com/post/132/what-are-covariance-and-contravariance) in their argument type. In other words, no function type is considered a subtype of `(...args: unknown[])=>any` because `unknown` is not considered a subtype of anything but `any` and `unknown`.

```ts
type myFunc = (arg: string) => boolean;
type test1 = myFunc extends (...args: any[]) => any ? true : false;
// true
type test2 = myFunc extends (...args: any[]) => unknown ? true : false;
// true
type test3 = myFunc extends (...args: unknown[]) => any ? true : false;
// false
```